src/SAML2/XML/samlp/Scoping.php

```diff
@@ -0,0 +1,166 @@
+
+use DOMElement;
+use SAML2\Constants;
+use SAML2\DOMDocumentFactory;
+use SAML2\Utils;
+use Webmozart\Assert\Assert;
+
+/**
+ * Class for handling SAML2 Scoping.
+ *
+ * @author Tim van Dijen, <tvdijen@gmail.com>
+ * @package simplesamlphp/saml2
+ */
+final class Scoping extends AbstractSamlpElement
+{
+    /** @var \SAML2\XML\samlp\IDPList|null */
+    protected $IDPList;
+
+    /** @var string[] */
+    protected $requesterId;
+
+    /** @var int|null */
+    protected $proxyCount;
+
+
+    /**
+     * Initialize a Scoping element.
+     *
+     * @param int|null $proxyCount
+     * @param \SAML2\XML\samlp\IDPList $idpList
+     * @param string[] $requesterId
+     */
+    public function __construct(?int $proxyCount = null, ?IDPList $idpList = null, array $requesterId = [])
+    {
+        $this->setProxyCount($proxyCount);
+        $this->setIDPList($idpList);
+        $this->setRequesterId($requesterId);
+    }
+
+
+    /**
+     * @return \SAML2\XML\samlp\IDPList|null
+     */
+    public function getIDPList(): ?IDPList
+    {
+        return $this->IDPList;
+    }
+
+
+    /**
+     * @param \SAML2\XML\samlp\IDPList|null $idpList
+     * @return void
+     */
+    private function setIDPList(?IDPList $idpList): void
+    {
+        $this->IDPList = $idpList;
+    }
+
+
+    /**
+     * @return string[]
+     */
+    public function getRequesterId(): array
+    {
+        return $this->requesterId;
+    }
+
+
+    /**
+     * @param string[] $requesterId
+     * @return void
+     */
+    private function setRequesterId(array $requesterId): void
+    {
+        $this->requesterId = $requesterId;
+    }
+
+
+    /**
+     * @return int|null
+     */
+    public function getProxyCount(): ?int
+    {
+        return $this->proxyCount;
+    }
+
+
+    /**
+     * @param int|null $proxyCount
+     * @return void
+     */
+    private function setProxyCount(?int $proxyCount): void
+    {
+        $this->proxyCount = $proxyCount;
+    }
+
+
+    /**
+     * Test if an object, at the state it's in, would produce an empty XML-element
+     *
+     * @return bool
+     */
+    public function isEmptyElement(): bool
+    {
+        return (
+            empty($this->proxyCount)
+            && empty($this->IDPList)
+            && empty($this->requesterId)
+        );
+    }
+
+
+    /**
+     * Convert XML into a Scoping-element
+     *
+     * @param \DOMElement $xml The XML element we should load
+     * @return \SAML2\XML\samlp\Scoping
+     * @throws \InvalidArgumentException if the qualified name of the supplied element is wrong
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        Assert::same($xml->localName, 'Scoping');
+        Assert::same($xml->namespaceURI, Scoping::NS);
+
+        $proxyCount = self::getAttribute($xml, 'ProxyCount', null);
+        $idpList = IDPList::getChildrenOfClass($xml);
+        $requesterId = Utils::extractStrings($xml, AbstractSamlpElement::NS, 'RequesterID');
```

Comment:
This makes me wonder if there is an actual way for `Utils::extractStrings()` to return an array with multiple strings, rather than one single string. Or if it even makes sense.

Reply:
Not sure what you're saying.. It'll return an array with multiple strings..

Reply:
I know, that's the signature, but I wonder if that's actually a legitimate use case. As far as I understand the code in `Utils`, the only way this can happen is when there are multiple elements with text content under the parent. It's just weird in my head.

Reply:
Yeah, Scoping can contain multiple RequesterIDs.. That's perfectly legit?
In most cases there will be only one result though..

Reply:
Uhm, I see! Thanks for the example! It feels kind of weird to handle those cases like this, but nevermind!